### PR TITLE
Add Providers to Blueprint

### PR DESCRIPTION
**Pull request by colincross**
Providers are a new concept for Blueprint, based on providers in Bazel:
https://docs.bazel.build/versions/master/skylark/rules.html#providers

Providers aim to simplify the interaction between modules by replacing
type asserting to arbitrary interfaces with requesting optional data
objects from modules.  This will also move Blueprint closer to supporting
incremental analysis by serializing the providers and only rerunning
the analysis phase on modules whose inputs have changed.

Change-Id: I39f5f78b372412a7dbf151ceccb3f917f6c874bf

### CODE REVIEW

File: module_ctx.go [hunk 1]
```diff
@@ -523,6 +542,29 @@ func (m *baseModuleContext) OtherModuleReverseDependencyVariantExists(name strin
 	return found != nil
 }
 
+// OtherModuleProvider returns the value, if any, for the provider for a module.  If the value for
```
Comment (gyias) [inline module_ctx.go:545]: This comment and the below one contain slightly different information than their interface counterparts'. Is this intended? If not, I'd just merge them into the interface and remove these so that we don't have to keep them in sync.

File: provider.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+package blueprint
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// This file implements Providers, modelled after Bazel
+// (https://docs.bazel.build/versions/master/skylark/rules.html#providers).
+// Each provider can be associated with a mutator, in which case the value for the provider for a
```
Comment (gyias) [inline provider.go:24]: What's the use case you're trying to support (or anti-patterns you're trying to disallow) by coupling providers with a mutator or gba? Should not specifying and thus allowing a provider to be set in any phase be an option?

File: provider_test.go [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+	if g, w := bModule.mutatorProviderValues, []string{"b", "c", "d", "d"}; !reflect.DeepEqual(g, w) {
+		t.Errorf("expected B.mutatorProviderValues %q, got %q", w, g)
+	}
+}
+
+type invalidProviderUsageMutatorInfo string
+type invalidProviderUsageGenerateBuildActionsInfo string
+
+var invalidProviderUsageMutatorInfoProvider = NewMutatorProvider(invalidProviderUsageMutatorInfo(""), "mutator_under_test")
+var invalidProviderusageGenerateBuildActionsInfoProvider = NewProvider(invalidProviderUsageGenerateBuildActionsInfo(""))
```
Comment (lizkammer) [inline provider_test.go:171]:
```suggestion
var invalidProviderUsageGenerateBuildActionsInfoProvider = NewProvider(invalidProviderUsageGenerateBuildActionsInfo(""))
```

and throughout?

**Review by lizkammer — APPROVED** [review]

File: provider.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+			panic(fmt.Sprintf("Can't set value of provider %s before GenerateBuildActions",
+				provider.typ))
+		} else if m.finishedGenerateBuildActions {
+			panic(fmt.Sprintf("Can't set value of provider %s after GenerateBuildActions",
+				provider.typ))
+		}
+	} else {
+		expectedMutator := c.providerMutators[provider.id]
+		if expectedMutator == nil {
+			panic(fmt.Sprintf("Can't set value of provider %s associated with unregistered mutator %s",
```
[the remark this answers is not in the document]
Comment (colincross) [reply provider.go:128]: Providers are registered globally at init, but unit tests use a context that register a subset of the mutators.  initProviders would panic for any test that didn't register all the mutators that have providers, here it will only panic if a test exercises a provider associated with an unregistered mutator.

File: module_ctx.go (hunk 1 above)
Comment (colincross) [reply module_ctx.go]: Done.

File: provider.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+// Once Go has generics the exampleValue parameter will not be necessary:
+// NewMutatorProvider(type T)(mutator string) ProviderKey(T)
+func NewMutatorProvider(exampleValue interface{}, mutator string) ProviderKey {
+	checkCalledFromInit()
+
+	typ := reflect.TypeOf(exampleValue)
+	zero := reflect.Zero(typ).Interface()
+
+	provider := &provider{
+		id:      len(providerRegistry),
```
[the remark this answers is not in the document]
Comment (colincross) [reply provider.go:86]: This has to be called from an init() function (verified by checkCalledFromInit() at the top of the function), and init() functions are never parallel in Go.

File: provider.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+			panic(fmt.Sprintf("Can't set value of provider %s after GenerateBuildActions",
+				provider.typ))
+		}
+	} else {
+		expectedMutator := c.providerMutators[provider.id]
+		if expectedMutator == nil {
+			panic(fmt.Sprintf("Can't set value of provider %s associated with unregistered mutator %s",
+				provider.typ, provider.mutator))
+		} else if c.finishedMutators[expectedMutator] || (c.startedMutator == expectedMutator && m.finishedMutator == expectedMutator) {
+			panic(fmt.Sprintf("Can't set value of provider %s after mutator %s",
```
[the remark this answers is not in the document]
Comment (colincross) [reply provider.go:131]: Done.

File: provider.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+				provider.typ, provider.mutator))
+		} else if c.startedMutator != expectedMutator || m.startedMutator != expectedMutator {
+			panic(fmt.Sprintf("Can't set value of provider %s before mutator %s",
```
[the remark this answers is not in the document]
Comment (colincross) [reply provider.go:134]: Done.

File: provider.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+	if provider.mutator == "" {
+		if !m.finishedGenerateBuildActions {
+			panic(fmt.Sprintf("Can't get value of provider %s before GenerateBuildActions",
+				provider.typ))
+		}
+	} else {
+		expectedMutator := c.providerMutators[provider.id]
+		if expectedMutator != nil &&
+			!c.finishedMutators[c.providerMutators[provider.id]] &&
+			(c.startedMutator != expectedMutator || m.finishedMutator != expectedMutator) {
```
[the remark this answers is not in the document]
Comment (colincross) [reply provider.go:173]: Added helper functions for both started and finished mutator checks.

File: provider_test.go (hunk 3 above)
Comment (colincross) [reply provider_test.go:171]: Done

File: provider.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lizkammer) [reply provider.go:86]: Good to know, thanks!

File: provider.go (hunk 2 above)
Comment (colincross) [reply provider.go:24]: Oops, accidentally marked this resolved.

I was worried about ordering.  Since it is valid for a provider to exist on some modules and not on others, and it is often very hard to follow the mutator ordering in Soong, I thought it would be common to read a provider from a module before the mutator that was supposed to set it, get the zero value, and be confused why it was always empty/unset.  This way the logic on whether or not a provider is set is guaranteed to be entirely contained within a single mutator.
